### DISCUSSION TITLE
ContractFilter mismatch at the EndpointDispatcher exception

### DISCUSSION THREAD

**Discussion by karthickprakash** [CoreWCF/CoreWCF#1466] [Q&A] [answered]
Hi,

I am working on converting SOAP services from BizTalk to .net core using CoreWCF. The client of the SOAP service is a web application using .net framework 4.8.

The client application has a proxy class generated form the WSDL of the BizTalk SOAP service. The Bindings and Endpoint URLS are configured in the app.config file. 

I monitored the SOAP requests using Fiddler. The SOAP request for the existing legacy service looks like

![image](https://github.com/user-attachments/assets/42f2ddc0-d0b3-46f8-8cb2-ed377c8c29b4)

I created a new SOAP service using CoreWCF and created a C# console application client. It uses proxy from the CoreWCF wsdl. The request for the CoreWCF service looks like

![image](https://github.com/user-attachments/assets/efd1eb2d-eb9c-4e15-b9ba-f5c2fd1fc9bb)

![image](https://github.com/user-attachments/assets/f4f2bfd9-3762-470f-8b4c-44de1ca76f36)

"P8_GetDocumentByID" is the Action name. I notice a difference between the soap body in the legacy request and the new request.

I tried to replace the legacy url with the new CoreWCF url in the app.config of the legacy client and I am getting the below exception

"The message with Action 'P8_GetDocumentByID' cannot be processed at the receiver, due to a ContractFilter mismatch at the EndpointDispatcher. This may be because of either a contract mismatch 
(mismatched Actions between sender and receiver) or a binding/security mismatch between the sender and the receiver.  
Check that sender and receiver have the same contract and the same binding (including security requirements, e.g. Message, Transport, None)."

Can you please let me know how to resolve the issue?

**Comment by mconnew** [comment]
What does the contract interface look like in the old codebase? Did you have a GetDocumentByIDRequest class before? One of two things is wrong. Either the method name used to be called GetDocumentByID (with no P8_ prefix) and it had a DocumentID parameter, and the return value was a single value, eg something like this:
```
[OperationContract(Action = "P8_GetDocumentByID")]
Task<string> GetDocumentByID(Guid DocumentID);
```
I presumed the return value is a string, but it could be something else, no way to know without seeing the response. The other alternative is that GetDocumentByIDRequest is attributed with `MessageContract`, and in the old codebase it has the property IsWrapped set to false, but on the new codebase that property is missing.  

You should be able to use the contract interface from your old codebase without change, and copy it to your service implementation and just switch the using from System.ServiceModel to CoreWCF.

**Reply by karthickprakash** [reply]
Thank you @mconnew for the reply. Adding [MessageContract(IsWrapped = false)] attribute to the class helped to make the request bodies consistent, but there is an issue with the namespaces.

![image](https://github.com/user-attachments/assets/240748fa-a30b-4c6d-ba32-690a9b0fda19)

### LEGACY REQUEST:

The legacy request does not have any namespace for the "DocumentID".

![image](https://github.com/user-attachments/assets/f0be1a51-08bc-453a-b534-362fe475c0f9)

### CORE WCF REQUEST:

But CoreWCF Request has a namespace before the "DocumentID".

![image](https://github.com/user-attachments/assets/27533ab0-c834-425a-bba0-2522118637d9)

I don't know if it is because of this, but the value for DocumentID is NULL in the CoreWCF service method.

![image](https://github.com/user-attachments/assets/b0821f24-c487-4be9-bfe2-289a3e201f5d)

Can you please help me with this?

Also, the Legacy response has a namespace included in the response

![image](https://github.com/user-attachments/assets/0e8abac6-d187-4d45-91c5-9022675a16da)

CoreWCF response has the same <a:> prefixed before the values

![image](https://github.com/user-attachments/assets/2570316a-028d-4604-933f-e7f453db45a9)

**Comment by mconnew** [comment, marked as answer]
I'm a little confused by your data types. GetDocumentByIDRequest is the type passed to the P8_GetDocumentByID method in your contract, but in the new code snippet, you have the MessageContract attribute added to the class P8_GetDocumentByIDRequest, which isn't used in your contract. It looks like you have one too many classes wrapping your request.   

Have you tried copying your contract from your legacy application? It should just work with a using change.  

And yes, that value is null because the namespace doesn't match. Here's what I think your request should look like:

```csharp
[ServiceContract(NameSpace = "FAI.Claims.FileManagementService")]
public interface IFileManagementServce
{
    [OperationContract(Action = "P8_GetDocumentByID")]
    Task<GetDocumentByIDResponse> GetDocumentByID(GetDocumentByIDRequest request);
}

[MessageContract(WrapperName = "GetDocumentByIDRequest", 
                 WrapperNamespace = "http://FileManagementService.FileManagement_Abridged")]
public class GetDocumentByIDRequest
{
    [MessageBodyMember(
      Name = "DocumentID",
      Namespace = "")]
    public Guid DocumentID { get; get; }
}
```
I don't know how you are having no namespace declared on the response. It almost looks like you are returning an XmlElement for the response. There are ways to do it with XmlElement attributes to say it's unqualified, but that seems an odd mix of technologies to specify in a WCF service.

**Reply by karthickprakash** [reply]
Thank you @mconnew  for the help. I was able to make things work by using the [MessageContract] and [MessageBodyMember] attributes.
